### PR TITLE
refactor(synthetic-shadow): move event patching to prototypes

### PR DESCRIPTION
**Pull request by caridy**
## Details

This is some preliminary work to make sure that we follow the proper patching order in synthetic shadow:

* remove `patchEvent()` method in favor of event prototype patches
* patch EventTarget (Node in IE11)
* patch window.add/removeEventListener to support retargeting
* move keys and owner keys into a separate utility that can be used from anywhere

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

### CODE REVIEW

File: packages/@lwc/synthetic-shadow/src/shared/node-ownership.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { defineProperty, isNull, isUndefined } from '@lwc/shared';
+import { parentNodeGetter } from '../env/node';
+
+// DO NOT CHANGE this:
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/shared/node-ownership.ts]: This comment is not valid anymore. @lwc/engine does not rely on these keys.

File: packages/@lwc/synthetic-shadow/src/shared/node-ownership.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+// these two values need to be in sync with engine
+const OwnerKey = '$$OwnerKey$$';
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/shared/node-ownership.ts]: For future reference, please add a comment here for each key type.
```suggestion
// Used as a back reference to identify the host element(owner) of a dom node
const OwnerKey = '$$OwnerKey$$';
```

File: packages/@lwc/synthetic-shadow/src/shared/node-ownership.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+const OwnKey = '$$OwnKey$$';
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/shared/node-ownership.ts]: How about calling this `HostKey`?

File: packages/@lwc/synthetic-shadow/src/index.ts [hunk 4]
```diff
@@ -15,15 +15,15 @@ import './env/mutation-observer';
...
-import './polyfills/document-shadow/main';
-import './polyfills/shadow-root/main';
-import './polyfills/event-listener/main';
 import './polyfills/click-event-composed/main'; // must come before event-composed
 import './polyfills/event-composed/main';
 import './polyfills/custom-event-composed/main';
 import './polyfills/focus-event-composed/main';
 import './polyfills/iframe-content-window/main';
 import './polyfills/mutation-observer/main';
+import './polyfills/event-listener/main';
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/index.ts]:
```suggestion
import './polyfills/event-listener/main'; // must come before ./faux-shadow/event-target
```

**Comment by ravijayaramappa** [comment]
This comment is related to https://github.com/salesforce/lwc/pull/1572 but since that is merged, adding it here.

IMHO addEventListener() in polyfills/event-listener/polyfill needs an assertion to make sure it is never invoked with a host element as context.

tl;dr;
With the latest master, `faux-shadow/event-target` is patching behavior for addEventListener called on a host element. `polyfill/event-listener` is providing behavior for all other elements. 
The switching between the code paths is subtly handled. event-target dereferences addEventListener from Node.target in the same module. event-listener accesses addEventListener from env/element. 
If the order of the polyfill and faux-shadow module changes, for example we refactored and moved code around, the wrong patch can be called on the host element. Adding a guard in polyfills/event-listener to make sure it is never called with a host would detect that.

File: packages/@lwc/synthetic-shadow/src/polyfills/event/polyfill.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            : originalCurrentTarget;
+    return retarget(currentTarget, composedPath);
+}
+
+function composedPathValue(this: Event): EventTarget[] {
+    const originalTarget: EventTarget = eventTargetGetter.call(this);
+    const originalCurrentTarget = eventCurrentTargetGetter.call(this);
+    // if the dispatch phase is done, the composedPath should be empty array
+    // TODO: in closed mode the path does not include shadowed objects from the currentTarget,
+    // should the following line call pathComposed() with this.currentTarget instead?
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/polyfills/event/polyfill.ts]: I didn't understand this comment. How does using `currentTarget` instead of `target` make a difference?

File: packages/@lwc/synthetic-shadow/src/polyfills/focus-event/polyfill.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import { retarget } from '../../3rdparty/polymer/retarget';
+
+const focusEventRelatedTargetGetter = getOwnPropertyDescriptor(
+    FocusEvent.prototype,
+    'relatedTarget'
+)!.get as () => EventTarget | null;
+
+defineProperty(FocusEvent.prototype, 'relatedTarget', {
+    get(this: Event): EventTarget | null | undefined {
+        const currentTarget = this.currentTarget;
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/polyfills/focus-event/polyfill.ts]:
```suggestion
        const currentTarget = eventCurrentTargetGetter.call(this);
```

File: packages/@lwc/synthetic-shadow/src/polyfills/mouse-event/polyfill.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import { retarget } from '../../3rdparty/polymer/retarget';
+
+const mouseEventRelatedTargetGetter = getOwnPropertyDescriptor(
+    FocusEvent.prototype,
+    'relatedTarget'
+)!.get as () => EventTarget | null;
+
+defineProperty(MouseEvent.prototype, 'relatedTarget', {
+    get(this: Event): EventTarget | null | undefined {
+        const currentTarget = this.currentTarget;
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/polyfills/mouse-event/polyfill.ts]:
```suggestion
        const currentTarget = eventCurrentTargetGetter.call(this);
```

File: packages/@lwc/synthetic-shadow/src/polyfills/mouse-event/polyfill.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        if (isNull(currentTarget)) {
+            // TODO: if currentTarget is null it is because relatedTarget is accessed
+            // in another turn, what should we do here? null seems safe, but probably
+            // not correct.
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/polyfills/mouse-event/polyfill.ts]: Should we document this limitation and remove this TODO?

File: packages/@lwc/synthetic-shadow/src/polyfills/focus-event/polyfill.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        if (isNull(currentTarget)) {
+            // TODO: if currentTarget is null it is because relatedTarget is accessed
+            // in another turn, what should we do here? null seems safe, but probably
+            // not correct.
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/polyfills/focus-event/polyfill.ts]: Should we document this limitation and remove this TODO?

File: packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+function isQualifyingEventTarget(currentTarget: Window, evt: Event): boolean {
+    const { composed } = evt;
+    const originalTarget = eventTargetGetter.call(evt);
+    if (originalTarget === currentTarget) {
+        // dispatched on the window directly
+        return true;
+    }
+    // if the event reaches the window by propagation is because it is bubbling,
+    // in which the the only it really depends on the composed value
+    return isTrue(composed) || !isNodeShadowed(originalTarget as Node);
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts]: Is this safe for the case where we have third party libraries listening for events dispatched from manually inserted elements inside LWC?

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts [hunk 11]
```diff
@@ -375,15 +170,6 @@ export function addShadowRootEventListener(
...
 ) {
-    if (process.env.NODE_ENV !== 'production') {
-        if (!isFunction(listener)) {
-            throw new TypeError(
-                `Invalid second argument for ShadowRoot.addEventListener() in ${toString(
-                    sr
-                )} for event "${type}". Expected an EventListener but received ${listener}.`
-            );
-        }
-    }
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts]: How come we are removing this guardrail?

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts [hunk 12]
```diff
@@ -320,42 +143,14 @@ function detachDOMListener(elm: Element, type: string, wrappedListener: WrappedL
...
 ) {
-    if (process.env.NODE_ENV !== 'production') {
-        if (!isFunction(listener)) {
-            throw new TypeError(
-                `Invalid second argument for Element.addEventListener() in ${toString(
-                    elm
-                )} for event "${type}". Expected an EventListener but received ${listener}.`
-            );
-        }
-    }
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts]: How come we are removing this guardrail?

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts [hunk 13]
```diff
@@ -178,68 +45,24 @@ function getWrappedShadowRootListener(
...
     let shadowRootWrappedListener = shadowRootEventListenerMap.get(listener);
     if (isUndefined(shadowRootWrappedListener)) {
         shadowRootWrappedListener = function(event: Event) {
-            // * if the event is dispatched directly on the host, it is not observable from root
-            // * if the event is dispatched in an element that does not belongs to the shadow and it is not composed,
-            //   it is not observable from the root
-            const { composed } = event as any;
+            // if the event is dispatched directly on the host, it is not observable from root,
+            // this is a limitation of the synthetic shadow polyfill
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts]: We briefly discussed offline that this should read more like, "if the event is dispatched directly on the shadow root, it is not observable from the root", but I don't think we support dispatching events on shadow roots, do we?

Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts]: thinking more about this... I think we can fix it by patching dispatchEvent on the shadowRoot, in which case, we know...

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts (hunk 12 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts]: because it is happening one layer on top... in the patch to `addEventListener`, which wraps every listener, and calls this method with the wrapped (always existing) function.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts (hunk 11 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts]: same as above... moved up one layer.

File: packages/@lwc/synthetic-shadow/src/polyfills/event-target/polyfill.ts [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+// otherwise it returns false, which means the listener will never be invoked.
+function isQualifyingEventTarget(currentTarget: EventTarget, evt: Event): boolean {
+    const { composed, bubbles } = evt;
+    if (!composed) {
+        const originalTarget = eventTargetGetter.call(evt);
+        if (bubbles) {
+            // bubbles true, composed false: only propagate up to the originalTarget's shadowRoot
+            if (originalTarget instanceof Node) {
+                if (isNodeShadowed(originalTarget)) {
+                    // if the original target is shadowed, every hub in the path to its shadowRoot
```
[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/polyfills/event-target/polyfill.ts]: replaced to `node`

File: packages/@lwc/synthetic-shadow/src/polyfills/event-target/polyfill.ts [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    options?: boolean | AddEventListenerOptions
+) {
+    if (listener == null) {
+        return; /* nullish */
+    }
+    const wrapperFn = getEventListenerWrapper(listener);
+    if (isHostElement(this)) {
+        addCustomElementEventListener(this, type, wrapperFn, options);
+    } else {
+        superAddEventListener.call(this, type, wrapperFn, options);
```
[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/polyfills/event-target/polyfill.ts]: well, this is the event target prototype... nothing weird here, it should support options.

in this case I think the issue is the .call instead of .apply, but that's more tricky to solve, I will look at it.

File: packages/@lwc/synthetic-shadow/src/polyfills/event/polyfill.ts [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+import { pathComposer } from '../../3rdparty/polymer/path-composer';
+import { retarget } from '../../3rdparty/polymer/retarget';
+import { getNodeOwnerKey } from '../../shared/node-ownership';
+import { getOwnerDocument } from '../../shared/utils';
+import { eventTargetGetter, eventCurrentTargetGetter } from '../../env/dom';
+import { getEventContext, EventListenerContext } from '../../shared/event-context';
+import { getShadowRoot } from '../../faux-shadow/shadow-root';
+
+function targetGetter(this: Event): EventTarget | null {
+    // currentTarget is always defined
```
[the remark this answers is not in the document]
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/polyfills/event/polyfill.ts]: you're right! removing... copy and pasted from another method.

File: packages/@lwc/synthetic-shadow/src/polyfills/mouse-event/polyfill.ts (hunk 7 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/polyfills/mouse-event/polyfill.ts]: intentionally using `this.currentTarget` here since it is used below for the retargeting logic. not sure what the implication is if we change it to the original current target.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts [hunk 17]
```diff
@@ -173,73 +40,40 @@ function getEventMap(elm: EventTarget): ListenerMap {
...
+
+export function setEventFromShadowRoot(event: Event) {
+    eventsComingFromShadowRoot.set(event, 1);
+}
 
 function getWrappedShadowRootListener(
     sr: SyntheticShadowRootInterface,
     listener: EventListener
 ): WrappedListener {
-    if (!isFunction(listener)) {
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts]: We are relaxing this condition. But we are not account for a listener which implements the [EventListener interface](https://developer.mozilla.org/en-US/docs/Web/API/EventListener).

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts [hunk 18]
```diff
@@ -173,73 +40,40 @@ function getEventMap(elm: EventTarget): ListenerMap {
...
-                    listener.call(sr, event);
-                }
+            // An event that is dispatched directly into the shadowRoot instance vs an event that is dispatched
+            // on the host instance directly are difficult to distinguish. We mark the event on the shadowRoot's
+            // dispatchEvent method, so we know it is a valid event for the shadowRoot.
+            if (target !== currentTarget || isEventComingFromShadowRoot(event)) {
+                listener.call(sr, event);
             }
         } as WrappedListener;
         shadowRootWrappedListener!.placement = EventListenerContext.SHADOW_ROOT_LISTENER;
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts]:
```suggestion
        shadowRootWrappedListener.placement = EventListenerContext.SHADOW_ROOT_LISTENER;
```

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts (hunk 17 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts]: good catch! I thought we were triggering this from another patch, but that's not the case... will look at solving this.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts [hunk 19]
```diff
@@ -105,11 +106,15 @@ defineProperties(HTMLSlotElement.prototype, {
...
             this: HTMLSlotElement,
             type: string,
-            listener: EventListener,
-            options?: boolean | AddEventListenerOptions
+            _listener: EventListener,
+            _options?: boolean | AddEventListenerOptions
         ) {
             // super.addEventListener - but that doesn't work with typescript
-            HTMLElement.prototype.addEventListener.call(this, type, listener, options);
+            HTMLElement.prototype.addEventListener.apply(this, ArraySlice.call(arguments) as [
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts]: Why not addEventListener from `../env/element`?

Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/slot.ts]: because that one is not patched, while `HTMLElement.prototype.addEventListener` is patched

File: packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts [hunk 20]
```diff
@@ -502,6 +509,17 @@ const NodePatchDescriptors = {
                 : this;
         },
     },
+    dispatchEvent: {
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts]: Should we lift [this restriction](https://github.com/salesforce/lwc/blob/4942cc44de98cae24862df4135f443ea13959555/packages/%40lwc/engine/src/framework/restrictions.ts#L272-L286) for dispatchEvent on shadow root?

File: packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+function getEventListenerWrapper(fnOrObj: EventListenerOrEventListenerObject): EventListener {
+    let wrapperFn: EventListener | undefined = (fnOrObj as any).$$lwcEventWrapper$$;
+    if (isUndefined(wrapperFn)) {
+        wrapperFn = (fnOrObj as any).$$lwcEventWrapper$$ = function(this: Window, e: Event) {
+            if (isQualifyingEventTarget(this, e)) {
+                if (typeof fnOrObj === 'function') {
+                    fnOrObj.call(this, e);
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts]: This snippet is repeated from polyfill/event-target/polyfill.ts, the only difference being the type specification of `this`. We should move this to a common util.

Also, if the same listener is used on window and document(or any extensions of EventTarget), the same cached wrapper will be used on both because the property names(.$$lwcEventWrapper$$) are the same. This is fine today because the wrapper logic is the same in both the polyfills. In the future if we change logic, then we might be accessing the wrong wrapper. I would prefer if we explicitly specify that the wrapper logic is the same by moving `getEventListenerWrapper` to a util or using a different property name to cache the wrapped listener.

File: packages/@lwc/synthetic-shadow/src/polyfills/mouse-event/polyfill.ts [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+ */
+import { defineProperty, isNull, getOwnPropertyDescriptor } from '@lwc/shared';
+import { pathComposer } from '../../3rdparty/polymer/path-composer';
+import { retarget } from '../../3rdparty/polymer/retarget';
+import { eventCurrentTargetGetter } from '../../env/dom';
+import { isNodeShadowed } from '../../shared/node-ownership';
+import { getOwnerDocument } from '../../shared/utils';
+
+const mouseEventRelatedTargetGetter = getOwnPropertyDescriptor(
+    FocusEvent.prototype,
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/polyfills/mouse-event/polyfill.ts]:
```suggestion
    MouseEvent.prototype,
```

File: packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts [hunk 23]
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { windowRemoveEventListener, windowAddEventListener } from '../../env/window';
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts]:
```suggestion
import { defineProperties, isUndefined, isTrue, ArraySlice } from '@lwc/shared';
import { windowRemoveEventListener, windowAddEventListener } from '../../env/window';
```

File: packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+// this method returns true if an event can be seen by a particular eventTarget,
+// otherwise it returns false, which means the listener will never be invoked.
+function isQualifyingEventTarget(currentTarget: Window, evt: Event): boolean {
+    const { composed } = evt;
+    const originalTarget = eventTargetGetter.call(evt);
+    if (originalTarget === currentTarget) {
+        // dispatched on the window directly
+        return true;
+    }
+    // if the event reaches the window by propagation is because it is bubbling,
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts]:
```suggestion
    // if the event reaches the window by propagation, it is because it is bubbling.
    // In which case, it really depends on the event's composed value
```

File: packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+    // in which the the only it really depends on the composed value
+    return isTrue(composed) || !isNodeShadowed(originalTarget as Node);
+}
+
+function getEventListenerWrapper(fnOrObj: EventListenerOrEventListenerObject): EventListener {
+    let wrapperFn: EventListener | undefined = (fnOrObj as any).$$lwcEventWrapper$$;
+    if (isUndefined(wrapperFn)) {
+        wrapperFn = (fnOrObj as any).$$lwcEventWrapper$$ = function(this: Window, e: Event) {
+            if (isQualifyingEventTarget(this, e)) {
+                if (typeof fnOrObj === 'function') {
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts]: We should retarget the event if it originated from inside the shadow.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts (hunk 20 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts]: excellent catch!

File: packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts (hunk 21 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts]: ok

File: packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts (hunk 23 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts]: I'm ok with this, first time I heard about this though.

File: packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts (hunk 21 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts]: I prefer to keep them separate... for now... it is not a lot of code.

as for the second part, lets just use a weakmap to be safe.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts (hunk 20 above)
Comment (caridy) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts]: fixed

File: packages/@lwc/engine/src/framework/restrictions.ts [hunk 26]
```diff
@@ -274,7 +274,6 @@ function getShadowRootRestrictionsDescriptors(
         getElementById: 0,
         getSelection: 0,
         elementsFromPoint: 0,
-        dispatchEvent: 0,
```
Comment (ravijayaramappa) [inline packages/@lwc/engine/src/framework/restrictions.ts]: @t-lau @manuel-jasso We should make sure locker allows dispatchEvent() on SecureTemplate.

File: packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts (hunk 10 above)
Comment (ravijayaramappa) [reply packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts]: I want to flag this as a concern too. This is potentially blocking some non-composed events bubbling from an aura components within an lwc container from reaching the window. 
The [current behavior](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/synthetic-shadow/src/polyfills/event-listener/polyfill.ts) is only deciding whether to patch or not to patch an event reaching the window. It never really blocks the event from reaching the window.

File: packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+        // dispatched on the window directly
+        return true;
+    }
+    const { composed } = evt;
+    // if the event reaches the window by propagation, it is because it is bubbling,
+    // in which the the only it really depends on the composed value
+    return isTrue(composed) || !isNodeShadowed(originalTarget as Node);
+}
+
+const EventListenerToWrapperMap: WeakMap<
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts]:
```suggestion
const eventListenerToWrapperMap: WeakMap<
```

File: packages/@lwc/synthetic-shadow/src/polyfills/mouse-event/polyfill.ts [hunk 28]
```diff
@@ -0,0 +1,36 @@
+/*
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/polyfills/mouse-event/polyfill.ts]: This polyfill is missing a readme

File: packages/@lwc/synthetic-shadow/src/polyfills/focus-event/polyfill.ts [hunk 29]
```diff
@@ -0,0 +1,36 @@
+/*
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/polyfills/focus-event/polyfill.ts]: Missing a README.md too

File: packages/@lwc/engine/src/framework/restrictions.ts (hunk 26 above)
Comment (ravijayaramappa) [reply packages/@lwc/engine/src/framework/restrictions.ts]: @arcadeteddy

**Comment by caridy** [comment]
@ravijayaramappa I'm out until Tue, can you help here?

File: packages/@lwc/engine/src/framework/restrictions.ts (hunk 26 above)
Comment (arcadeteddy) [reply packages/@lwc/engine/src/framework/restrictions.ts]: We currently don't allow it. https://developer.salesforce.com/docs/component-library/documentation/lwc/lwc.security_locker_global_access I think back then the synthetic ShadowRoot didn't support it either... If it doesn't cause any security concerns, we can whitelist it.

File: packages/@lwc/synthetic-shadow/src/polyfills/event-target/polyfill.ts [hunk 30]
```diff
@@ -0,0 +1,155 @@
...
+            }
+        };
+        EventListenerToWrapperMap.set(fnOrObj, wrapperFn);
+    }
+    return wrapperFn;
+}
+
+// These methods are usually from EventTarget.prototype, but that's not available in IE11, the next best thing
+// is Node.prototype, which is an EventTarget as well.
+const {
```
Comment (ravijayaramappa) [inline packages/@lwc/synthetic-shadow/src/polyfills/event-target/polyfill.ts]: I believe this can be extracted in `../../env/node` and reuse it here. Node.prototype.addEventListener is not patched before this point.

File: packages/@lwc/engine/src/framework/restrictions.ts (hunk 26 above)
Comment (caridy) [reply packages/@lwc/engine/src/framework/restrictions.ts]: yes, let's open a task to track that @arcadeteddy

**Comment by animmakayala** [comment]
W-8015735

File: packages/@lwc/engine/src/framework/restrictions.ts (hunk 26 above)
Comment (caridy) [reply packages/@lwc/engine/src/framework/restrictions.ts:277]: @arcadeteddy can you confirm that this is now allowed?

File: packages/@lwc/synthetic-shadow/src/polyfills/event/polyfill.ts (hunk 5 above)
Comment (ekashida) [reply packages/@lwc/synthetic-shadow/src/polyfills/event/polyfill.ts:53]: Yeah, the `pathComposer` algorithm isn't taking into account the mode so it seems that we should be passing in the `currentTarget` or the `target` based on that.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts [hunk 31]
```diff
@@ -172,74 +39,48 @@ function getEventMap(elm: EventTarget): ListenerMap {
...
-                );
-
-                if (
-                    isChildNode(rootNode as HTMLElement, currentTarget as Node) ||
-                    (composed === false && rootNode === currentTarget)
-                ) {
+            // An event that is dispatched directly into the shadowRoot instance vs an event that is dispatched
+            // on the host instance directly are difficult to distinguish. We mark the event on the shadowRoot's
+            // dispatchEvent method, so we know it is a valid event for the shadowRoot.
+            if (target !== currentTarget || isEventComingFromShadowRoot(event)) {
```
Comment (ekashida) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts:69]: Looks like we are no longer filtering out events that are not composed with this change...

**Comment by ekashida** [comment]
This PR was broken down into the following:
https://github.com/salesforce/lwc/pull/2131
https://github.com/salesforce/lwc/pull/2117
https://github.com/salesforce/lwc/pull/2139
https://github.com/salesforce/lwc/pull/2111
https://github.com/salesforce/lwc/issues/2121
